fe2-android/app/src/test/java/com/github/dedis/student20_pop/model/data/LAORepositoryTest.java

```diff
@@ -37,121 +43,201 @@
 
-    answerSubscribe.subscribe(testObserverSubscribe);
+    // Check the correct DataSource is being used
+    Mockito.verify(remoteDataSource, Mockito.times(1)).sendMessage(Mockito.any());
 
-    testScheduler.advanceTimeBy(950, TimeUnit.MILLISECONDS);
-    testObserverSubscribe.assertNotComplete();
+    // Check there is no answer before the response delay time
+    testScheduler.advanceTimeTo(RESPONSE_DELAY - 1, TimeUnit.MILLISECONDS);
+    testObserver.assertNotComplete();
 
-    testScheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
-    testObserverSubscribe.assertComplete().assertNoErrors()
-        .assertValue(r -> r.getId() == 42);
+    // Check the subscribe result is ready at response delay time
+    // and verify the id of the result corresponds to the expected id
+    testScheduler.advanceTimeTo(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+    testObserver.assertComplete().assertNoErrors()
+        .assertValue(r -> r.getId() == REQUEST_ID);
   }
 
   @Test
   public void testSendUnsubscribe() {
-    Single<Answer> answerUnsubscribe = repository.sendUnsubscribe("/root/");
+    // Send an unsubscribe request and subscribe to the answer
+    Single<Answer> answerUnsubscribe = repository.sendUnsubscribe(CHANNEL);
+    answerUnsubscribe.subscribe(testObserver);
+
+    // Check the correct DataSource is being used
     Mockito.verify(remoteDataSource, Mockito.times(1)).sendMessage(Mockito.any());
 
-    TestObserver<Answer> testObserverUnsubscribe = TestObserver.create();
+    // Check there is no answer before the response delay time
+    testScheduler.advanceTimeTo(RESPONSE_DELAY - 1, TimeUnit.MILLISECONDS);
+    testObserver.assertNotComplete();
 
-    answerUnsubscribe.subscribe(testObserverUnsubscribe);
+    // Check the unsubscribe result is ready at response delay time
+    // and verify the id of the result corresponds to the expected id
+    testScheduler.advanceTimeTo(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+    testObserver.assertComplete().assertNoErrors()
+        .assertValue(r -> r.getId() == REQUEST_ID);
+  }
+
+  @Test
+  public void testGetAllLaos() {
+    // Does not use LAOLocalDataSource
+
+    // Create the LAO and the subscriber to test from
+    Lao lao = new Lao(LAO_CHANNEL);
+    TestObserver<List<Lao>> subscriber = TestObserver.create();
 
-    testScheduler.advanceTimeBy(950, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertNotComplete();
+    // Subscribe to a LAO and wait for the request to finish
+    repository.sendSubscribe(LAO_CHANNEL);
+    testScheduler.advanceTimeBy(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
 
-    testScheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertComplete().assertNoErrors()
-        .assertValue(r -> r.getId() == 42);
+    repository.getAllLaos().subscribe(subscriber);
+
+    // Check the LAO is present in the laoById map of LAORepository
+    List<Lao> repositoryLaoList = (List<Lao>) subscriber.getEvents().get(0).get(0);
```

Comment:
what's `get(0).get(0)` doing here? 😅 
it's not exactly obvious.
it might not hurt to actually assign intermediate steps to named variables.

Reply:
`getEvents` returns a list of `(values, errors, completion-notifications)`, the idea is to get the first value received from `getAllLaos` after the subscription. I will change this to make it more readable